analyzers/src/SonarAnalyzer.VisualBasic/Rules/UseReturnStatement.cs

```diff
@@ -0,0 +1,90 @@
+ */
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class UseReturnStatement : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S5944";
+        private const string MessageFormat = "{0}";
+        private const string UseReturnStatementMessage = "Use a 'Return' statement; assigning returned values to function names is obsolete.";
+        private const string DontUseImplicitMessage = "Do not make use of the implicit return value.";
+
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var method = (MethodStatementSyntax)((MethodBlockSyntax)c.Node).BlockStatement;
+                new IdentifierWalker(c, method.Identifier.ValueText).SafeVisit(c.Node);
+            },
+            SyntaxKind.FunctionBlock);
+
+        private class IdentifierWalker : VisualBasicSyntaxWalker
+        {
+            private readonly SyntaxNodeAnalysisContext context;
+            private readonly string name;
+
+            public IdentifierWalker(SyntaxNodeAnalysisContext context, string name)
+            {
+                this.context = context;
+                this.name = name;
+            }
+
+            public override void VisitIdentifierName(IdentifierNameSyntax node)
+            {
+                if (IsImplicitReturnValue(node))
+                {
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, node.GetLocation(),
+                        IsAssignmentStatement(node)
+                        ? UseReturnStatementMessage
+                        : DontUseImplicitMessage));
+                }
+            }
+
+            private bool IsImplicitReturnValue(IdentifierNameSyntax node) =>
+                name.Equals(node.Identifier.ValueText, StringComparison.InvariantCultureIgnoreCase)
+                && !IsExcluded(node);
+
+            private static bool IsExcluded(SyntaxNode node) =>
+                node.Parent is InvocationExpressionSyntax
+                || node.Parent is MemberAccessExpressionSyntax
+                || node.Parent is NamedFieldInitializerSyntax
+                || node.Ancestors().OfType<ImplementsClauseSyntax>().Any();
```

Comment:
This is looks like a hacky solution. We're inside a walker so we can stop walking on implements clause
```suggestion
```
```
            public override void VisitImplementsClause(ImplementsClauseSyntax node)
            {
                // Do not process Identifier inside Implements InterfaceName.MethodNameIdentifier
            }
```

Reply:
All 3 statements better are solved by skipping those nodes. I updated the implementation.

Reply:
But that would change the behavior of the code. Because `Invocation(NoncompliantMethodNameAsArgument)` would not be visited like this.

Can you revert it back to `IsExcluded` for those 3 parents? And remove the `Skip` as it would not be needed.